### DISCUSSION TITLE
Built the C++ library  Error

### DISCUSSION THREAD

**Discussion by DancyHuang** [apache/incubator-graphar#597] [Q&A]
When i  built the C++ library of GraphAr：

>>> cmake ..
-- The C compiler identification is GNU 9.4.0
-- The CXX compiler identification is GNU 9.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'Release' as none was specified.
-- [graphar] will build in type: Release
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Checking for module 'thrift'
--   Found thrift, version 0.13.0
CMake Warning at CMakeLists.txt:200 (message):
  apache-arrow is built without ORC extension, ORC related functionalities
  will be disabled.

-- Found Protobuf: /usr/lib/x86_64-linux-gnu/libprotobuf.so;-pthread (found version "3.6.1") 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/fh/newgraphar/graphar/cpp/build-release

>>> make -j
Scanning dependencies of target graphar
[  5%] Building CXX object CMakeFiles/graphar.dir/src/graphar/arrow/chunk_reader.cc.o
[ 11%] Building CXX object CMakeFiles/graphar.dir/src/graphar/arrow/chunk_writer.cc.o
[ 17%] Building CXX object CMakeFiles/graphar.dir/src/graphar/chunk_info_reader.cc.o
[ 23%] Building CXX object CMakeFiles/graphar.dir/src/graphar/chunk_info_writer.cc.o
[ 29%] Building CXX object CMakeFiles/graphar.dir/src/graphar/expression.cc.o
[ 35%] Building CXX object CMakeFiles/graphar.dir/src/graphar/filesystem.cc.o
/home/fh/newgraphar/graphar/cpp/src/graphar/filesystem.cc: In member function ‘std::shared_ptr<arrow::dataset::FileFormat> graphar::FileSystem::GetFileFormat(graphar::FileType) const’:
/home/fh/newgraphar/graphar/cpp/src/graphar/filesystem.cc:91:47: error: could not convert ‘std::make_shared(_Args&& ...) [with _Tp = arrow::dataset::CsvFileFormat; _Args = {}]()’ from ‘shared_ptr<arrow::dataset::CsvFileFormat>’ to ‘shared_ptr<arrow::dataset::FileFormat>’
   91 |     return std::make_shared<ds::CsvFileFormat>();
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
      |                                               |
      |                                               shared_ptr<arrow::dataset::CsvFileFormat>
/home/fh/newgraphar/graphar/cpp/src/graphar/filesystem.cc:93:51: error: could not convert ‘std::make_shared(_Args&& ...) [with _Tp = arrow::dataset::ParquetFileFormat; _Args = {}]()’ from ‘shared_ptr<arrow::dataset::ParquetFileFormat>’ to ‘shared_ptr<arrow::dataset::FileFormat>’
   93 |     return std::make_shared<ds::ParquetFileFormat>();
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
      |                                                   |
      |                                                   shared_ptr<arrow::dataset::ParquetFileFormat>
/home/fh/newgraphar/graphar/cpp/src/graphar/filesystem.cc:95:48: error: could not convert ‘std::make_shared(_Args&& ...) [with _Tp = arrow::dataset::JsonFileFormat; _Args = {}]()’ from ‘shared_ptr<arrow::dataset::JsonFileFormat>’ to ‘shared_ptr<arrow::dataset::FileFormat>’
   95 |     return std::make_shared<ds::JsonFileFormat>();
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
      |                                                |
      |                                                shared_ptr<arrow::dataset::JsonFileFormat>
In file included from /usr/include/c++/9/bits/shared_ptr_base.h:56,
                 from /usr/include/c++/9/bits/shared_ptr.h:52,
                 from /usr/include/c++/9/memory:81,
                 from /usr/local/include/arrow/array/array_base.h:22,
                 from /usr/local/include/arrow/array.h:41,
                 from /usr/local/include/arrow/api.h:22,
                 from /home/fh/newgraphar/graphar/cpp/src/graphar/filesystem.cc:23:
/usr/include/c++/9/ext/aligned_buffer.h: In instantiation of ‘struct __gnu_cxx::__aligned_buffer<arrow::dataset::CsvFileFormat>’:
/usr/include/c++/9/bits/shared_ptr_base.h:535:35:   required from ‘class std::_Sp_counted_ptr_inplace<arrow::dataset::CsvFileFormat, std::allocator<arrow::dataset::CsvFileFormat>, __gnu_cxx::_S_atomic>::_Impl’
/usr/include/c++/9/bits/shared_ptr_base.h:596:13:   required from ‘class std::_Sp_counted_ptr_inplace<arrow::dataset::CsvFileFormat, std::allocator<arrow::dataset::CsvFileFormat>, __gnu_cxx::_S_atomic>’
/usr/include/c++/9/bits/shared_ptr_base.h:676:43:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = arrow::dataset::CsvFileFormat; _Alloc = std::allocator<arrow::dataset::CsvFileFormat>; _Args = {}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr_base.h:1344:71:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<arrow::dataset::CsvFileFormat>; _Args = {}; _Tp = arrow::dataset::CsvFileFormat; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr.h:359:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<arrow::dataset::CsvFileFormat>; _Args = {}; _Tp = arrow::dataset::CsvFileFormat]’
/usr/include/c++/9/bits/shared_ptr.h:701:14:   required from ‘std::shared_ptr<_Tp> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = arrow::dataset::CsvFileFormat; _Alloc = std::allocator<arrow::dataset::CsvFileFormat>; _Args = {}]’
/usr/include/c++/9/bits/shared_ptr.h:717:39:   required from ‘std::shared_ptr<_Tp> std::make_shared(_Args&& ...) [with _Tp = arrow::dataset::CsvFileFormat; _Args = {}]’
/home/fh/newgraphar/graphar/cpp/src/graphar/filesystem.cc:91:48:   required from here
/usr/include/c++/9/ext/aligned_buffer.h:91:28: error: invalid application of ‘sizeof’ to incomplete type ‘arrow::dataset::CsvFileFormat’
   91 |     : std::aligned_storage<sizeof(_Tp), __alignof__(_Tp)>
      |                            ^~~~~~~~~~~
/usr/include/c++/9/ext/aligned_buffer.h:91:28: error: invalid application of ‘sizeof’ to incomplete type ‘arrow::dataset::CsvFileFormat’
/usr/include/c++/9/ext/aligned_buffer.h: In instantiation of ‘struct __gnu_cxx::__aligned_buffer<arrow::dataset::ParquetFileFormat>’:
/usr/include/c++/9/bits/shared_ptr_base.h:535:35:   required from ‘class std::_Sp_counted_ptr_inplace<arrow::dataset::ParquetFileFormat, std::allocator<arrow::dataset::ParquetFileFormat>, __gnu_cxx::_S_atomic>::_Impl’
/usr/include/c++/9/bits/shared_ptr_base.h:596:13:   required from ‘class std::_Sp_counted_ptr_inplace<arrow::dataset::ParquetFileFormat, std::allocator<arrow::dataset::ParquetFileFormat>, __gnu_cxx::_S_atomic>’
/usr/include/c++/9/bits/shared_ptr_base.h:676:43:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = arrow::dataset::ParquetFileFormat; _Alloc = std::allocator<arrow::dataset::ParquetFileFormat>; _Args = {}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr_base.h:1344:71:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<arrow::dataset::ParquetFileFormat>; _Args = {}; _Tp = arrow::dataset::ParquetFileFormat; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr.h:359:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<arrow::dataset::ParquetFileFormat>; _Args = {}; _Tp = arrow::dataset::ParquetFileFormat]’
/usr/include/c++/9/bits/shared_ptr.h:701:14:   required from ‘std::shared_ptr<_Tp> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = arrow::dataset::ParquetFileFormat; _Alloc = std::allocator<arrow::dataset::ParquetFileFormat>; _Args = {}]’
/usr/include/c++/9/bits/shared_ptr.h:717:39:   required from ‘std::shared_ptr<_Tp> std::make_shared(_Args&& ...) [with _Tp = arrow::dataset::ParquetFileFormat; _Args = {}]’
/home/fh/newgraphar/graphar/cpp/src/graphar/filesystem.cc:93:52:   required from here
/usr/include/c++/9/ext/aligned_buffer.h:91:28: error: invalid application of ‘sizeof’ to incomplete type ‘arrow::dataset::ParquetFileFormat’
/usr/include/c++/9/ext/aligned_buffer.h:91:28: error: invalid application of ‘sizeof’ to incomplete type ‘arrow::dataset::ParquetFileFormat’
/usr/include/c++/9/ext/aligned_buffer.h: In instantiation of ‘struct __gnu_cxx::__aligned_buffer<arrow::dataset::JsonFileFormat>’:
/usr/include/c++/9/bits/shared_ptr_base.h:535:35:   required from ‘class std::_Sp_counted_ptr_inplace<arrow::dataset::JsonFileFormat, std::allocator<arrow::dataset::JsonFileFormat>, __gnu_cxx::_S_atomic>::_Impl’
/usr/include/c++/9/bits/shared_ptr_base.h:596:13:   required from ‘class std::_Sp_counted_ptr_inplace<arrow::dataset::JsonFileFormat, std::allocator<arrow::dataset::JsonFileFormat>, __gnu_cxx::_S_atomic>’
/usr/include/c++/9/bits/shared_ptr_base.h:676:43:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = arrow::dataset::JsonFileFormat; _Alloc = std::allocator<arrow::dataset::JsonFileFormat>; _Args = {}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr_base.h:1344:71:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<arrow::dataset::JsonFileFormat>; _Args = {}; _Tp = arrow::dataset::JsonFileFormat; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr.h:359:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<arrow::dataset::JsonFileFormat>; _Args = {}; _Tp = arrow::dataset::JsonFileFormat]’
/usr/include/c++/9/bits/shared_ptr.h:701:14:   required from ‘std::shared_ptr<_Tp> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = arrow::dataset::JsonFileFormat; _Alloc = std::allocator<arrow::dataset::JsonFileFormat>; _Args = {}]’
/usr/include/c++/9/bits/shared_ptr.h:717:39:   required from ‘std::shared_ptr<_Tp> std::make_shared(_Args&& ...) [with _Tp = arrow::dataset::JsonFileFormat; _Args = {}]’
/home/fh/newgraphar/graphar/cpp/src/graphar/filesystem.cc:95:49:   required from here
/usr/include/c++/9/ext/aligned_buffer.h:91:28: error: invalid application of ‘sizeof’ to incomplete type ‘arrow::dataset::JsonFileFormat’
/usr/include/c++/9/ext/aligned_buffer.h:91:28: error: invalid application of ‘sizeof’ to incomplete type ‘arrow::dataset::JsonFileFormat’
make[2]: *** [CMakeFiles/graphar.dir/build.make:128：CMakeFiles/graphar.dir/src/graphar/filesystem.cc.o] 错误 1
make[1]: *** [CMakeFiles/Makefile2:106：CMakeFiles/graphar.dir/all] 错误 2
make: *** [Makefile:130：all] 错误 2

**Comment by acezen** [comment]
Hi, @DancyHuang , what is the version of the apache arrow in your environment?

**Comment by DancyHuang** [comment]
the version of the apache arrow17.0.0




At 2024-08-20 10:58:16, "Weibin Zeng" ***@***.***> wrote:

Hi, @DancyHuang , what is the version of the apache arrow in your environment?

**Reply by acezen** [reply]
Can you give your environment message, like
- OS system
- GraphAr version or commit

**Reply by DancyHuang** [reply]
OS system: Ubuntu 20.04.6 LTS
GraphAr version :  master

**Reply by acezen** [reply]
Hi, @DancyHuang, I can not reproduce the problem in Ubuntu 20.04, can you install the dependencies with [setup-ubuntu.sh](https://github.com/apache/incubator-graphar/blob/main/dev/release/setup-ubuntu.sh) and try to compile again?
